server/shared/sse/index.js

```diff
@@ -37,46 +41,57 @@ class SSEConnection extends EventEmitter {
     return this._res.req;
   }
 
+  get socket() {
+    return this._res.socket;
+  }
+
   get timeout() {
-    const connectionTimeout = parseInt(this.request.header('connection-timeout'), 0);
+    const connectionTimeout = parseInt(this.request.header('connection-timeout'), 10);
     const timeout = connectionTimeout || SSE_DEFAULT_TIMEOUT;
     return timeout * (1 - SSE_TIMEOUT_MARGIN);
   }
 
   initialize() {
-    const { socket } = this.request;
-    socket.setTimeout(0);
-    socket.setNoDelay(true);
-    socket.setKeepAlive(true);
+    this.socket.setTimeout(0);
+    this.socket.setNoDelay(true);
+    this.socket.setKeepAlive(true);
     // Gracefully handle termination.
     this.request.once('close', () => this.close());
     // Set event stream headers.
-    this._res.status(200).type('text/event-stream').set(SSE_HEADERS);
+    this._res.writeHead(200, SSE_HEADERS);
     this._res.flushHeaders();
-    // Start stream.
-    this._writeln(':ok');
-    this._writeln();
-    this.send('connection_initialized', { sseId: this.id });
     // Setup heartbeat interval.
-    if (!this.timeout) return;
-    this._heartbeat = setInterval(() => {
-      this._writeln(':ping');
-      this._writeln();
-    }, this.timeout);
+    if (this.timeout > 0) {
+      this._heartbeat = setInterval(() => this.write(':ping'), this.timeout);
+    }
+    // Start stream.
+    return this.write(':ok');
+  }
+
+  write(payload = '') {
+    return this._res.write(`${payload}\n\n`);
   }
 
   send(event, data = '') {
-    this._lastEventId += 1;
-    this._writeln(`id = ${this._lastEventId}`);
-    this._writeln(`event: ${event}`);
-    const lines = JSON.stringify(data).split(/\r?\n/g);
-    lines.forEach(line => this._writeln(`data: ${line}`));
-    this._writeln();
+    const id = this._lastEventId += 1;
+    this.emit('data', { id, event, data });
+    const json = JSON.stringify(data);
+    const payload = [
+      `id: ${id}`,
+      `event: ${event}`,
+      `data: ${json}`
+    ].join('\n');
+    this.write(payload);
+    if (hasProp(this.request.query, 'debug')) {
+      this.debug({ id, type: event, data });
+    }
```

Comment:
In order to support client-side event logging, every event is re-emitted as a `message` with the actual event payload.